### ISSUE TITLE
PR #1789: GraphAttentionNetwork training is nondeterministic (flaky in isolation, not a CI flake)

### ISSUE THREAD

**Issue by ooples** [ooples/AiDotNet#1860]
## Summary
`GraphAttentionNetwork` (GAT) training tests are **nondeterministic even in ISOLATION** — this is a real training-determinism bug, NOT a CI-parallel flake (reclassified out of #1857).

## Evidence
- Original test (crypto init + static param cache), run twice SOLO: run 1 = 2 failed, run 2 = 0 failed.
- Pinning a fixed thread-local init seed (`LayerInitializationSeedScope`) did NOT stabilize it: seeded, it still failed a *varying* subset (3 failed one run, 1 failed the next). A fixed init seed makes weight init deterministic, so the remaining nondeterminism is in **training**, not init.
- SpiralNet (same seed treatment) became fully stable (3/3), so the seed mechanism works — GAT has a distinct source.

## Failing tests (shard: ModelFamily - NeuralNetworks G-L)
`GraphAttentionNetworkTests`: Clone_AfterTraining, DifferentInputs_AfterTraining, ForwardPass_ShouldBeFinite_AfterTraining, GradientFlow_ShouldBeNonZeroAndFinite, OptimizerStep_ParamL2_DoesNotExplode, MoreData (the exact set varies run-to-run).

## Likely root cause
Non-deterministic parallel reductions (float non-associativity under `Parallel.For`) in the GAT attention/neighbor-aggregation, and/or an attention-dropout RNG path not covered by the per-layer init seed. Compare against the `DeterministicReductions` pattern used to de-flake other MoreData families, and ensure GAT's attention softmax/aggregation and dropout are seeded/deterministic. Fix the determinism in the model/training path; do NOT weaken the tests.

## Repro (run a few times — it is intermittent)
```
dotnet test tests/AiDotNet.Tests/AiDotNetTests.csproj --framework net10.0 -c Release --no-build \
  --filter "FullyQualifiedName~GraphAttentionNetworkTests"
```

## Note
The prior `static _savedParams` cache in the test masks nothing here — the nondeterminism is in the training path. A fix should make GAT training reproducible under a fixed seed, after which pinning the seed (as SpiralNet now does) will also close the CI-parallel exposure.

**Comment by ooples** [comment]
## Refined diagnosis (the flake is in the BACKWARD, and survives every standard determinism control)

Narrowed it down considerably:

- **Not the forward.** `Predict_ShouldBeDeterministic` passes; every failing test is `*_AfterTraining` (Clone/DifferentInputs/ForwardPass/GradientFlow/OptimizerStep/MoreData). So the nondeterminism is in **training (the backward/gradient), not the forward**.
- **Not the input.** The test input uses `ModelTestHelpers.CreateSeededRandom()` = fixed seed 42.
- **Not weight init / attention-dropout RNG.** `GraphAttentionLayer` seeds `_random` (both init AND the 0.6 attention-dropout mask) from `RandomSeed`, which `LayerBase` wires from `LayerInitializationSeedScope`. I pinned it via `AmbientFallbackSeed = 1337` around `CreateNetwork`; GAT was STILL flaky across processes (runs gave 1, 0, then 3 failures with the SAME seed). A fixed seed that leaves results non-reproducible ⇒ the remaining nondeterminism is not RNG.
- **Not BLAS / GPU.** The harness has `AiDotNetEngine.SetDeterministicMode(true)` (ModuleInitializer) + `BlasProvider.SetDeterministicMode(true)` (openblas threads=1) + `AIDOTNET_DISABLE_GPU=1`. SpiralNet (same seed treatment, also has Dropout) became fully deterministic (3/3), so the engine's common reductions ARE deterministic — GAT hits a distinct path.

**Conclusion:** a non-deterministic floating-point reduction in GAT's specific gradient path (the attention matmul / softmax-over-neighbors / ReduceSum-over-heads backward) that is NOT covered by `openblas_set_num_threads(1)` — i.e. a non-BLAS parallel reduction. This is a determinism gap at the Engine/tape level for those ops, not a one-line model fix.

**Recommended next step:** run GAT training twice in-process and diff per-parameter gradients after one backward to pin the exact op whose gradient differs, then make that reduction deterministic (or extend SetDeterministicMode to cover the non-BLAS reduction path). Reverted my seed exploration to keep the tree clean; the seed is a correct partial improvement (removes the init+dropout nondeterminism) but does not green GAT alone.

**Comment by ooples** [comment]
Still open after checking PR #1789 — no training-determinism fix for GAT exists on that branch. Worth flagging one near-miss: `bf18c4ebf` fixes a hand-rolled init loop that used `RandomHelper.CreateSecureRandom()` and therefore *bypassed* `LayerInitializationSeedScope`. That is a real seed-reproducibility bug, but it does **not** close this issue — the evidence here already records that pinning a fixed init seed did not stabilize GAT, which is what localized the nondeterminism to training rather than init.

**Comment by ooples** [comment]
### Re-audit against master (`origin/master` @ `a22d7e78bf` (2026-09-05))

**Why this correction exists.** The `Triage note (PR #1789 sweep)` comment on this issue was produced by checking the **#1789 branch**, which was cut on **2026-07-05**. That sweep ran on **2026-08-14**, the day before #1789 merged. Any fix that reached `master` in that 5.5-week window *outside* #1789 was invisible to it, so those verdicts are false-negative-prone. Re-verified below against master itself.

**Verdict: UNCERTAIN - passing is not the same as fixed**

Re-checked on master. `GraphAttentionNetworkTests` carries no `HeavyTimeout` trait and is running in the default gate, and the latest executed run (`33938630584`) is **115/115 shards passed** with GAT among them.

That is *not* sufficient evidence to close. This issue documents **nondeterminism** - a flaky test passing on one run is the expected behaviour of an unfixed flake, and the last comment records that pinning a fixed init seed did not stabilize it.

Recommend a repeat-run probe (the same test executed N times in isolation) before any close decision. Kept open deliberately.

---
*Posted by an automated re-audit of all 45 open issues against master on 2026-09-05. Every figure above was measured on the tree or read from CI run artifacts; none is carried over from a previous comment.*
